File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Services/CertStore.cs
using System.Text.Json.Nodes;
using Azure.AI.Vision.Face.DeviceAttestation.Crypto;
using Azure.AI.Vision.Face.DeviceAttestation.Storage;

namespace Azure.AI.Vision.Face.DeviceAttestation.Services;

/// <summary>Result of saving a certificate: its thumbprint and whether it was newly created.</summary>
internal sealed record SaveCertificateResult(string Thumbprint, bool IsNew);

/// <summary>
/// Certificate store, keyed by the SHA-256 thumbprint of the cert DER.
/// Persistence, key namespace, and TTL are delegated to the injected
/// <see cref="IClusterStore"/>; this layer keeps the domain concerns
/// (thumbprint computation, domain events, metadata merge).
/// </summary>
internal static class CertStore
{
    /// <summary>
    /// Save a certificate: create it on first sight, or bump its
    /// <c>lastVerifiedAt</c> on subsequent calls. Returns the thumbprint + a new
    /// flag, or null on error.
    /// </summary>
    public static async Task<SaveCertificateResult?> SaveCertificateAsync(
        AttestationContext ctx,
        string clientId,
        string system,
        string publicCert,
        JsonObject? metadata)
    {
        var thumbprint = CertUtils.ComputeCertThumbprint(publicCert);
        if (thumbprint is null)
        {
            ctx.Logger.TrackEvent("CertStore.SaveFail", new Dictionary<string, object?>
            {
                ["reason"] = "THUMBPRINT_COMPUTE_FAIL",
                ["clientId"] = clientId,
                ["system"] = system,
            });
            return null;
        }

        var existing = await ctx.Store.GetCertificateAsync(thumbprint);
        if (existing is not null)
        {
            if (existing.Value.ClientId != clientId || existing.Value.System != system) return null;
            existing.Value.LastVerifiedAt = IsoTime.Now();
            if (await ctx.Store.UpdateCertificateAsync(thumbprint, existing.Version, existing.Value) != UpdateResult.Applied)
            {
                return null;
            }
            ctx.Logger.TrackEvent("CertStore.Updated", new Dictionary<string, object?>
            {
                ["thumbprint"] = thumbprint,
                ["clientId"] = clientId,
                ["system"] = system,
                ["isNew"] = false,
            });
            return new SaveCertificateResult(thumbprint, false);
        }

        var now = IsoTime.Now();
        var certData = new CertificateData
        {
            ClientId = clientId,
            System = system,
            Thumbprint = thumbprint,
            PublicCert = publicCert,
            CreatedAt = now,
            LastVerifiedAt = now,
            Metadata = metadata,
        };
        if (!await ctx.Store.SetCertificateAsync(thumbprint, certData))
        {
            return null;
        }
        ctx.Logger.TrackEvent("CertStore.Created", new Dictionary<string, object?>
        {
            ["thumbprint"] = thumbprint,
            ["clientId"] = clientId,
            ["system"] = system,
            ["isNew"] = true,
        });
        return new SaveCertificateResult(thumbprint, true);
    }

    /// <summary>Load a certificate record by thumbprint.</summary>
    public static Task<Snapshot<CertificateData>?> GetCertificateAsync(AttestationContext ctx, string thumbprint)
        => ctx.Store.GetCertificateAsync(thumbprint);

    /// <summary>
    /// Merge fields into a cert record's metadata and bump <c>lastVerifiedAt</c>,
    /// preserving the existing TTL. Used to advance the assertion sign-count.
    /// Returns false if the key is gone/expired.
    /// </summary>
    public static async Task<bool> UpdateCertificateMetadataAsync(
        AttestationContext ctx,
        string thumbprint,
        JsonObject partialMetadata,
        Snapshot<CertificateData> snapshot)
    {
        var existing = snapshot.Value;

        existing.LastVerifiedAt = IsoTime.Now();
        var merged = existing.Metadata ?? new JsonObject();
        foreach (var kv in partialMetadata)
        {
            merged[kv.Key] = kv.Value?.DeepClone();
        }
        existing.Metadata = merged;

        return await ctx.Store.UpdateCertificateAsync(thumbprint, snapshot.Version, existing) == UpdateResult.Applied;
    }
}

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/services/CertStore.java
package com.azure.ai.vision.face.deviceattestation.services;

import java.util.HashMap;
import java.util.Map;

import com.azure.ai.vision.face.deviceattestation.AttestationContext;
import com.azure.ai.vision.face.deviceattestation.IsoTime;
import com.azure.ai.vision.face.deviceattestation.Maps;
import com.azure.ai.vision.face.deviceattestation.crypto.CertUtils;
import com.azure.ai.vision.face.deviceattestation.store.CertificateData;
import com.azure.ai.vision.face.deviceattestation.store.Snapshot;
import com.azure.ai.vision.face.deviceattestation.store.UpdateResult;

/**
 * Certificate store, keyed by the SHA-256 thumbprint of the cert DER.
 * Persistence, key namespace, and TTL are delegated to the injected store; this
 * layer keeps the domain concerns (thumbprint computation, domain events,
 * metadata merge).
 */
public final class CertStore {

    /**
     * Save a certificate: create it on first sight, or bump its {@code lastVerifiedAt}
     * on subsequent calls. Returns the thumbprint + a new flag, or null on error.
     */
    public static SaveCertificateResult saveCertificate(
            AttestationContext ctx,
            String clientId,
            String system,
            String publicCert,
            Map<String, Object> metadata) {
        String thumbprint = CertUtils.computeCertThumbprint(publicCert);
        if (thumbprint == null) {
            ctx.logger.trackEvent("CertStore.SaveFail",
                    Maps.of("reason", "THUMBPRINT_COMPUTE_FAIL", "clientId", clientId, "system", system), null);
            return null;
        }

        Snapshot<CertificateData> existing = ctx.store.getCertificate(thumbprint);
        if (existing != null) {
            if (!clientId.equals(existing.value().clientId) || !system.equals(existing.value().system)) return null;
            existing.value().lastVerifiedAt = IsoTime.now();
            if (ctx.store.updateCertificate(thumbprint, existing.version(), existing.value()) != UpdateResult.APPLIED) {
                return null;
            }
            ctx.logger.trackEvent("CertStore.Updated",
                    Maps.of("thumbprint", thumbprint, "clientId", clientId, "system", system, "isNew", false), null);
            return new SaveCertificateResult(thumbprint, false);
        }

        String now = IsoTime.now();
        CertificateData certData = new CertificateData();
        certData.clientId = clientId;
        certData.system = system;
        certData.thumbprint = thumbprint;
        certData.publicCert = publicCert;
        certData.createdAt = now;
        certData.lastVerifiedAt = now;
        certData.metadata = metadata;
        if (!ctx.store.setCertificate(thumbprint, certData)) {
            return null;
        }
        ctx.logger.trackEvent("CertStore.Created",
                Maps.of("thumbprint", thumbprint, "clientId", clientId, "system", system, "isNew", true), null);
        return new SaveCertificateResult(thumbprint, true);
    }

    /** Load a certificate record by thumbprint. */
    public static Snapshot<CertificateData> getCertificate(AttestationContext ctx, String thumbprint) {
        return ctx.store.getCertificate(thumbprint);
    }

    /**
     * Merge fields into a cert record's metadata and bump {@code lastVerifiedAt},
     * preserving the existing TTL. Used to advance the assertion sign-count.
     * Returns false if the key is gone/expired.
     */
    public static boolean updateCertificateMetadata(
            AttestationContext ctx,
            String thumbprint,
            Map<String, Object> partialMetadata,
            Snapshot<CertificateData> snapshot) {
        CertificateData existing = snapshot.value();

        existing.lastVerifiedAt = IsoTime.now();
        Map<String, Object> merged = existing.metadata != null ? existing.metadata : new HashMap<>();
        merged.putAll(partialMetadata);
        existing.metadata = merged;

        return ctx.store.updateCertificate(thumbprint, snapshot.version(), existing) == UpdateResult.APPLIED;
    }

    private CertStore() {
    }
}
